### PR TITLE
refactoring current metrics exporter implementation

### PR DESCRIPTION
**Pull request by abhilashshetty04**
This PR refactors current metrics exporter implementation by making it more modular, It makes adding more resource metrics easier.

This PR includes helm changes as well.

### CODE REVIEW

File: metrics-exporter/src/bin/io_engine/cache/mod.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use std::sync::Mutex;
+use tokio::time::sleep;
+use tracing::error;
+
+static CACHE: OnceCell<Mutex<Cache>> = OnceCell::new();
+
+/// Trait to be implemented by all Resource structs stored in Cache.
+pub trait ResourceOps {
```
Comment (tiagolobocastro) [inline metrics-exporter/src/bin/io_engine/cache/mod.rs:17]: Why are these public? and many more across the file

File: metrics-exporter/src/bin/io_engine/cache/mod.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    tokio::spawn(async move {
+        store_resource_data(client).await;
+    });
+}
+
+/// To store pools related data in cache.
+pub async fn store_resource_data(client: GrpcClient) {
+    loop {
+        if pool::store_pool_info_data(client.clone()).await.is_err() {
+            error!("error updating pool cache")
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply metrics-exporter/src/bin/io_engine/cache/mod.rs:83]: hmm a store method should also be part of the trait?

File: metrics-exporter/src/bin/io_engine/cache/pool.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        Err(error) => {
+            error!(%error, "Error while getting cache resource");
+            return Err(());
+        }
+    };
+    let pools_cache = cache.deref_mut();
+    match pools {
+        // set pools in the cache
+        Ok(pools) => {
+            debug!("Updated io_engine cache with latest metrics");
```
Comment (tiagolobocastro) [inline metrics-exporter/src/bin/io_engine/cache/pool.rs:38]:
```suggestion
            trace!("Updated pool cache with latest metrics");
```

File: metrics-exporter/src/bin/io_engine/main.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        .expect("gRPC client not initialized");
+
+    store_data(client).await;
+    info!("starting actix app");
+    let app = move || {
+        actix_web::App::new()
+            .wrap(middleware::Logger::default())
+            .configure(metric_route)
+    };
+    info!("starting endpoint");
```
[the remark this answers is not in the document]
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/main.rs:96]: Will remove

File: metrics-exporter/src/bin/io_engine/cache/mod.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    pub fn pool_mut(&mut self) -> &mut Pools {
+        &mut self.data.pools
+    }
+}
+
+/// Wrapper over all the data that has to be stored in cache.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Data {
+    /// Contains Pool Capacity and state data.
+    pools: Pools,
```
[the remark this answers is not in the document]
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/cache/mod.rs:49]: You mean trait name can be changed?

File: metrics-exporter/src/bin/io_engine/collector/pool.rs [hunk 6]
```diff
@@ -1,47 +1,43 @@
...
+    /// Initialize all the metrics to be defined for pools capacity collector.
     pub fn new() -> Self {
-        let pool_total_size_opts = Opts::new("total_size_bytes", "Total size of the pool in bytes")
-            .subsystem("disk_pool")
-            .variable_labels(vec!["node".to_string(), "name".to_string()]);
-        let pool_used_size_opts = Opts::new("used_size_bytes", "Used size of the pool in bytes")
-            .subsystem("disk_pool")
-            .variable_labels(vec!["node".to_string(), "name".to_string()]);
+        let pool_total_size_opts =
+            Opts::new("total_size_bytes", "Total size of the io_engine in bytes")
```
[the remark this answers is not in the document]
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/collector/pool.rs:28]: Yes it happened when refatoring using ide option. WIll change

File: metrics-exporter/src/bin/io_engine/cache/mod.rs (hunk 1 above)
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/cache/mod.rs]: This is not public now. Its used within module. For others i have made crate-public.

File: metrics-exporter/src/bin/io_engine/cache/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/cache/mod.rs]: As per the discussion. We will merge this bit as is now. Will revisit later.

File: metrics-exporter/src/bin/io_engine/client/grpc_client.rs [hunk 7]
```diff
@@ -1,11 +1,10 @@
-use std::time::Duration;
-
-use crate::pool::{client::ApiVersion, error::ExporterError};
+use crate::{error::ExporterError, get_node_name, get_pod_ip, ApiVersion};
 use actix_web::http::Uri;
 use rpc::io_engine::IoEngineClientV0;
```
Comment (Abhinandan-Purkait) [inline metrics-exporter/src/bin/io_engine/client/grpc_client.rs:3]: Move this import up with the crate to seperate it from non-our crates

**Review by Abhinandan-Purkait — APPROVED** [review]

File: metrics-exporter/src/bin/io_engine/cache/mod.rs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use std::sync::Mutex;
+use tokio::time::sleep;
+static CACHE: OnceCell<Mutex<Cache>> = OnceCell::new();
+
+/// Trait to be implemented by all Resource structs stored in Cache.
+trait ResourceOps {
+    type ResourceVec;
```
[the remark this answers is not in the document]
Comment (abhilashshetty04) [reply metrics-exporter/src/bin/io_engine/cache/mod.rs:16]: Actually its always going to be vector of resource.

**Review by sinhaashish — APPROVED** [review]
LGTM

**Review by tiagolobocastro — APPROVED** [review]
Please amend your commit to refactor

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**